From: Misael Lopez Cruz <misael.lopez@ti.com>
Date: Sat, 29 Jan 2011 18:12:19 +0000 (-0600)
Subject: ASoC: Add dai link enable/disable
X-Git-Url: http://git.omapzoom.org/?p=kernel%2Fomap.git;a=commitdiff_plain;h=031e63480b6c1b9e5f3d07ec3cac033bc0391963

ASoC: Add dai link enable/disable

Add enable/disable callbacks for DAI links to setup link
preconditions (like clocks) during PCM startup, and clear
them after PCM shutdown.

Change-Id: I6707c911efffcbdb5d1a13d1dab6d57d42622780
Signed-off-by: Misael Lopez Cruz <misael.lopez@ti.com>
---

Index: kernel/include/sound/soc.h
===================================================================
--- kernel.orig/include/sound/soc.h	2011-04-29 12:22:28.279605395 +0900
+++ kernel/include/sound/soc.h	2011-04-29 13:04:54.880898026 +0900
@@ -613,6 +613,10 @@
 	int (*be_hw_params_fixup)(struct snd_soc_pcm_runtime *rtd,
 			struct snd_pcm_hw_params *params);
 
+	/* preconditions/postconditions for dai link */
+	int (*enable)(struct snd_soc_pcm_runtime *rtd);
+	void (*disable)(struct snd_soc_pcm_runtime *rtd);
+
 	/* machine stream operations */
 	struct snd_soc_ops *ops;
 };
Index: kernel/sound/soc/soc-core.c
===================================================================
--- kernel.orig/sound/soc/soc-core.c	2011-04-29 12:22:34.559605593 +0900
+++ kernel/sound/soc/soc-core.c	2011-04-29 13:04:54.880898026 +0900
@@ -576,6 +576,15 @@
 	if (rtd->dai_link->no_host_mode == SND_SOC_DAI_LINK_NO_HOST)
 		snd_soc_set_runtime_hwparams(substream, &no_host_hardware);
 
+	if (rtd->dai_link->enable) {
+		ret = rtd->dai_link->enable(rtd);
+		if (ret < 0) {
+			printk(KERN_ERR "asoc: can't enable dai link %s\n",
+				rtd->dai_link->name);
+			goto link_err;
+		}
+	}
+
 	/* startup the audio subsystem */
 	if (cpu_dai->driver->ops->startup) {
 		ret = cpu_dai->driver->ops->startup(substream, cpu_dai);
@@ -726,6 +735,9 @@
 	if (cpu_dai->driver->ops->shutdown)
 		cpu_dai->driver->ops->shutdown(substream, cpu_dai);
 out:
+	if (rtd->dai_link->disable)
+		rtd->dai_link->disable(rtd);
+link_err:
 	if (rtd->dai_link->dynamic)
 		snd_soc_put_backend_dais(substream);
 
@@ -860,6 +872,9 @@
 					SND_SOC_DAPM_STREAM_STOP);
 	}
 
+	if (rtd->dai_link->disable)
+		rtd->dai_link->disable(rtd);
+
 	if (rtd->dai_link->dynamic)
 		snd_soc_put_backend_dais(substream);
 	if(codec_dai->driver->playback.stream_name 
